torch/csrc/jit/runtime/profiling_record.cpp

```diff
@@ -144,35 +144,43 @@ void ProfilingRecord::insertShapeProfile(Node* n, size_t offset) {
   auto pno = pn->addOutput();
   pn->ty_(attr::profiled_type, TensorType::get());
   pno->setType(TensorType::get());
-  std::function<void(Stack&)> shape_profiler = [this, pno](Stack& stack) {
+  std::function<void(Stack&)> shape_profiler = [this, pn, pno](Stack& stack) {
     int64_t frame_id = 0;
     pop(stack, frame_id);
     IValue v;
     pop(stack, v);
+    std::lock_guard<std::mutex> lock(this->mutex_);
+    if (profiling_count_ <= 0) {
+      return;
+    }
+
     if (v.isTensor()) {
-      std::lock_guard<std::mutex> lock(this->mutex_);
-      auto& profiled_types = profiled_types_per_frame_[frame_id];
       auto& t = v.toTensor();
-      if (t.defined()) {
-        auto pttp = tensorTypeInCurrentExecutionContext(t);
+      auto new_tensor_type =
+          (t.defined() ? tensorTypeInCurrentExecutionContext(t)
+                       : TensorType::get()->withUndefined());
+      GRAPH_DEBUG(
+          "In run ",
+          frame_id,
+          " annotating %",
+          pno->debugName(),
+          " with ",
+          *new_tensor_type);
+
+      if (pn->hasRun()) {
+        auto existing_tensor_type =
+            pn->ty(attr::profiled_type)->expect<TensorType>();
         GRAPH_DEBUG(
-            "In run ",
-            frame_id,
-            " annotating %",
+            "Existing type for %",
             pno->debugName(),
-            " with ",
-            *pttp);
-        if (profiled_types.count(pno) == 0) {
-          profiled_types.insert({pno, pttp});
-        } else {
-          auto type = profiled_types.at(pno);
-          GRAPH_DEBUG("Existing type for %", pno->debugName(), " ", *type);
-          pttp = type->merge(*pttp);
-          GRAPH_DEBUG("Result for %", pno->debugName(), " ", *pttp);
-          profiled_types[pno] = pttp;
-        }
+            ": ",
+            *existing_tensor_type);
+        auto merged_type = new_tensor_type->merge(*existing_tensor_type);
+        GRAPH_DEBUG("Merged type for %", pno->debugName(), ": ", *merged_type);
+        pn->ty_(attr::profiled_type, merged_type);
```

Comment:
save a refcount bump here. would be slightly faster still to inline the call when the GRAPH_DEBUG won't fire, but we probably don't need to go there until profiling data says we do.
`         pn->ty_(attr::profiled_type, std::move(merged_type));`